### DISCUSSION TITLE
Petal Hero - rhythm game compatible with Frets on Fire for MicroPython-powered flow3r badge (ESP32-S3)

### DISCUSSION THREAD

**Discussion by dos1** [micropython/micropython#13076] [Show and tell]
[![mpv-shot0001](https://github.com/micropython/micropython/assets/126204/7de711cb-af7a-43a7-8f6e-f1733676f4ec)](https://social.librem.one/@dos/111478238181935805)

Hello! I've recently made a game using MicroPython and wanted to show it off a bit: it's called **[Petal Hero](https://git.flow3r.garden/dos/PetalHero)** and it's a Frets on Fire/Guitar Hero-like game made for the flow3r badge's small circular display and capacitive touch input around it. It's compatible with songs for several popular rhythm games (MIDI format).

[flow3r](https://flow3r.garden) is the [Chaos Communication Camp 2023](https://events.ccc.de/camp/2023/infos/index.html)'s badge. It's based on ESP32-S3 N16R8 and uses MicroPython as part of its [st3m](https://git.flow3r.garden/flow3r/flow3r-firmware) framework for writing Python apps. It provides APIs to handle the hardware and bindings to [ctx](https://ctx.graphics/) for vector rendering and [bl00mbox](https://gitlab.com/moon2embeddedaudio/bl00mbox) for audio synthesis; st3m can also play media files, in this case MP3. All the rest in the game is implemented as Python code (it even uses some snippets copied directly from the original Frets on Fire codebase :smile:)

Petal Hero can also be launched in the [flow3r badge simulator](https://git.flow3r.garden/flow3r/flow3r-firmware/-/tree/main/sim?ref_type=heads), although its interface makes it rather hard to actually play :wink:

I'm impressed in how usable the end result turned out to be! When I started the project I considered it an experiment that's quite likely to fail, but it ended up becoming a decent game. With recent MicroPython's garbage collection improvements for PSRAM on ESP32 it even forgives plenty of memory allocation sins, both in st3m and in the game code. GC stalls are still noticeable, but they don't cause that much trouble in the gameplay. Overall it was a pretty fun project :grin: 

See it in action on video: https://social.librem.one/@dos/111478238181935805

**Comment by bidrohini** [comment]
Not only fun, but the project looks very beautiful. Also compact. Thanks for sharing. If you ever consider making a students' trainer board using ESP32-S3, you can make one like [this](https://www.pcbway.com/project/shareproject/ESP32_S3_students_training_board_d03e9933.html):
